calico_kubernetes/calico_kubernetes_cni.py

```diff
@@ -307,6 +352,337 @@ def _get_container_pid(container_id):
+        _log.info('Adding tag %s' % ns_tag)
+        profile.tags.add(ns_tag)
+    else:
+        _log.warning('Namespace tag cannot be generated')
+
+    # Create tags from labels
+    labels = _get_metadata(namespace, pod_name, 'labels')
+    if labels:
+        for k, v in labels.iteritems():
+            tag = _label_to_tag(k, v)
+            _log.info('Adding tag ' + tag)
+            profile.tags.add(tag)
+
+    _datastore_client.profile_update_tags(profile)
+
+    _log.info('Finished applying tags.')
+
+
+def _generate_rules(container_id, namespace, pod_name, profile_name):
+    """Generate rules based on user set policy and namespace
+
+    This function returns a Rules datastore object.
+    Users can set policy using the environment variable DEFAULT_POLICY.
+    The default policy is allow all incoming and outgoing traffic.
+    If pod belongs to kube-system namespace, always allow all incoming and
+       outgoing traffic.
+
+    :return rules: a Rules object with attached inbound and outbound Rule objects
+    """
+    _log.info("Generating rules for pod %s", pod_name)
+
+    inbound_rules=[]
+    outbound_rules=[]
+
+    # kube-system services need to be accessed by all namespaces
+    if namespace == "kube-system" :
+        _log.info("Pod %s belongs to the kube-system namespace - "
+                  "Creating rules to allow all inbound and outbound traffic", pod_name)
+        inbound_rules.append(Rule(action="allow"))
+        outbound_rules.append(Rule(action="allow"))
+    elif namespace and DEFAULT_POLICY == 'ns_isolation':
+        _log.info("Creating rules to only allow traffic from namespace %s and "
+                  "allow all outgoing traffic", namespace)
+        ns_tag = _get_namespace_tag(namespace)
+        inbound_rules = Rule(action="allow", src_tag=ns_tag)
+        outbound_rules = Rule(action="allow")
+    else:
+        _log.info("Creating rules to allow all incoming and outgoing traffic")
+        inbound_rules = Rule(action="allow")
+        outbound_rules = Rule(action="allow")
+
+    rules = Rules(id=profile_name,
+                  inbound_rules=[inbound_rules],
+                  outbound_rules=[outbound_rules])
+
+    return rules
+
+
+def _get_annotations_rules(namespace, pod_name):
+    """Get rules from a pod's annotations
+
+    :param namepsace: Namespace that pod belongs to
+    :param pod_name: Name of pod
+    :return: List of inbound rules
+    """
+    _log.info("Looking for rules on the pod's annotations")
+
+    inbound_rules = []
+    outbound_rules = []
+    annotations = _get_metadata(namespace, pod_name, "annotations")
+
+    if annotations and POLICY_ANNOTATION_KEY in annotations:
+        _log.info("Getting policy rules from annotation of pod %s", pod_name)
+
+        # Remove Default Rule (Allow Namespace)
+        rules = annotations[POLICY_ANNOTATION_KEY]
+
+        # Rules separated by semicolons
+        for rule in rules.split(";"):
+            args = rule.split(" ")
```

Comment:
I think it's cleaner to say `args = rule.split()`. The default split will turn `a  b` into ['a', 'b'], so you don't have to do your filter() below (and it'll handle other whitespace too, like tabs). That should prevent straggling empty list members, both inside your list and at the ends.